File: skills/loi/scripts/backends/webhook.py
"""Webhook notify backend — POSTs JSON events to an HTTP endpoint.

Config:
    notify_url      The full URL to POST to (required)
    auth_token_env  Name of env var holding a bearer token (optional)

Payload shape follows the LOI event schema:
    {
      "event": "proposal.updated",
      "repo": "my-repo",
      "path": "docs/index/auth/_root.md",
      "timestamp": "2026-04-10T12:00:00Z",
      "summary": "...",
      "governance": {"health": "normal", "security": "normal"}
    }
"""

from __future__ import annotations

import json
import os
import urllib.request
from datetime import datetime, timezone


class WebhookBackend:
    """POST LOI events as JSON to a generic HTTP webhook endpoint."""

    def __init__(self, url: str, *, token_env: str | None = None) -> None:
        if not url:
            raise ValueError("WebhookBackend requires a non-empty notify_url")
        self.url = url
        self.token_env = token_env

    def _headers(self) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if self.token_env:
            token = os.environ.get(self.token_env, "")
            if token:
                headers["Authorization"] = f"Bearer {token}"
        return headers

    def send(self, event_type: str, payload: dict) -> None:
        event = {
            "event": event_type,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            **payload,
        }
        data = json.dumps(event).encode("utf-8")
        req = urllib.request.Request(
            self.url,
            data=data,
            headers=self._headers(),
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=10) as resp:
            if resp.status == 200:
                print(f"[LOI] Webhook event sent: {event_type}")
            else:
                print(f"[LOI] Webhook returned HTTP {resp.status} for event {event_type}")

File: internal/notify/webhook.go
package notify

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"
)

// webhookBackend POSTs JSON-encoded NotifyEvents to an HTTP URL.
// An optional bearer token is read from an environment variable at send time.
type webhookBackend struct {
	url      string
	tokenEnv string
	client   *http.Client
}

// newWebhookBackend returns a webhookBackend targeting url.
// If tokenEnv is non-empty the value of os.Getenv(tokenEnv) is sent as a
// Bearer token on every request.
func newWebhookBackend(url, tokenEnv string) *webhookBackend {
	return &webhookBackend{
		url:      url,
		tokenEnv: tokenEnv,
		client:   &http.Client{Timeout: 10 * time.Second},
	}
}

// Send marshals e as JSON and POSTs it to the configured URL.
func (b *webhookBackend) Send(e NotifyEvent) error {
	data, err := json.Marshal(e)
	if err != nil {
		return fmt.Errorf("notify/webhook: marshal: %w", err)
	}

	req, err := http.NewRequest(http.MethodPost, b.url, bytes.NewReader(data))
	if err != nil {
		return fmt.Errorf("notify/webhook: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	if b.tokenEnv != "" {
		if token := os.Getenv(b.tokenEnv); token != "" {
			req.Header.Set("Authorization", "Bearer "+token)
		}
	}

	resp, err := b.client.Do(req)
	if err != nil {
		return fmt.Errorf("notify/webhook: POST %s: %w", b.url, err)
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("notify/webhook: POST %s: unexpected status %d", b.url, resp.StatusCode)
	}
	return nil
}
